source/common/protobuf/utility.cc

```diff
@@ -556,17 +560,57 @@ void redact(Protobuf::Message* message, bool ancestor_is_sensitive) {
       // `MessageUtil::redact()` for more context.
       ENVOY_LOG_MISC(warn, "Could not redact udpa::type::v1::TypedStruct with unknown type URL {}",
                      typed_struct->type_url());
-      return;
+      return false;
     }
 
     MessageUtil::jsonConvert(typed_struct->value(), ProtobufMessage::getNullValidationVisitor(),
                              *typed_message);
     redact(typed_message.get(), ancestor_is_sensitive);
     MessageUtil::jsonConvert(*typed_message, *(typed_struct->mutable_value()));
+    return true;
+  }
+  return false;
+}
+
+// `DataSource`, which is annotated as `sensitive` in `TlsCertificate` and `TlsSessionTicketKeys`
+// requires special handling.
+template <typename DataSource>
+bool redactDataSource(Protobuf::Message* message, bool ancestor_is_sensitive) {
```

Comment:
Not super excited about this pattern, but if it's limited to just `DataSource`, I think we can live with it.

Reply:
Same. I'm hoping @PiotrSikora lets me off the hook for this special case :)

Reply:
I guess the point of doing that is: by the time an SDS related issue occurs, if indeed it's some subtle difference of not set  by SDS sever vs redacted by Envoy, and we can't tell by looking at log or config_dump. That can be tricky and time consuming debugging case. Although given currently in TlsCert part we have cert chain retained, thus we can reasonably mitigated this concern.